torch/csrc/jit/register_prim_ops.cpp

```diff
@@ -1009,14 +1019,236 @@ Operator(                                                                      \
 });
 
 
+// checking one of size & scale_factor is set
+// if scale_factor is a double list check that it's len == dim
+// reference: _check_size_scale_factor in torch/nn/functional.py
+void _check_size_factor(size_t dim, const IValue& size, IValue scale_factor) {
+  if (size.isNone() && scale_factor.isNone()) {
+    throw std::runtime_error("either size or scale_factor should be defined");
+  }
+  if (!size.isNone() && !scale_factor.isNone()) {
+    throw std::runtime_error("only one of size or scale_factor should be defined");
+  }
+  if (scale_factor.isDoubleList()) {
+    auto scale_len = scale_factor.toDoubleListRef().size();
+    if (scale_len != dim) {
+      std::stringstream str;
+      str << "scale_factor shape must match input shape. Input is " << dim
+        << "D, scale_factor size is " << scale_len;
+      throw std::runtime_error("only one of size or scale_factor should be defined");
+    }
+  }
+}
+
+// reference: _output_size in torch/nn/functional.py
+// size can be none, int or intlist
+// scale_factors can be none, float, or floatlist
+std::vector<int64_t> _output_size(at::Tensor input, size_t dim, IValue size, IValue scale_factors) {
+  if (!size.isNone()) {
+    if (size.isInt()) {
+      std::vector<int64_t> repeated(dim, size.toInt());
+      return repeated;
+    } else {
+      return size.toIntListRef();
+    }
+  }
+  std::vector<double> scale_repeated;
+  if (scale_factors.isDouble()) {
+    scale_repeated = std::vector<double>(dim, scale_factors.toDouble());
+  } else {
+    scale_repeated = scale_factors.toDoubleListRef();
+  }
+  std::vector<int64_t> ret;
+  for (size_t i = 0; i < dim; ++i) {
+    ret.push_back(std::floor(input.size(i + 2) * scale_repeated[i]));
+  }
+  return ret;
+}
+
+// reference: interpolate in torch/nn/functional.py
+// size can be none, int or intlist
+// scale_factors can be none, float, or floatlist
+at::Tensor interpolate(at::Tensor input, IValue size, IValue scale_factors,
```

Comment:
Is there a reason why this isn't written as a native function?

Reply:
If it were a native function the function would have to have 4 different variants, one for each combination of size & scale_factor. Plus doing it this way makes the logic in c++ and the existing python function 1-1. 

It is kind of annoying case. It might be better to move it to builtin_functions.cpp, but it would require having 4 variants and/or adding some overhead to support them, plus there are other features that are not yet implemented in builtin_functions that this would require.